### ISSUE TITLE
Pyright rejects valid and acepts invalid code

### ISSUE THREAD

**Issue by ethframe** [microsoft/pyright#2103]
**Describe the bug**
Pyright seems to mix up type variables bindings in some cases and infers wrong type for an expression.

**To Reproduce**
```python
from typing import Generic, Tuple, TypeVar

from typing_extensions import Protocol

T = TypeVar('T')
U = TypeVar('U')


class A(Protocol[T]):
    def a(self) -> T:
        ...

    def b(self, v: 'A[U]') -> 'A[Tuple[T, U]]':
        ...


class B(Generic[T]):
    def __init__(self, t: T):
        self._t = t

    def a(self) -> T:
        return self._t

    def b(self, v: A[U]) -> A[Tuple[T, U]]:
        return B((self._t, v.a()))
```
produces following error
```
Expression of type "B[tuple[T@B, U@b]]" cannot be assigned to return type "A[Tuple[T@B, U@b]]"
  TypeVar "T@A" is invariant
    Type "tuple[tuple[T@B, U@b], U@b]" cannot be assigned to type "Tuple[T@B, U@b]"
      Tuple entry 1 is incorrect type
        Type "tuple[T@B, U@b]" cannot be assigned to type "T@B"
      Tuple entry 1 is incorrect type
        Type "tuple[T@B, U@b]" cannot be assigned to type "T@B"
```

**Expected behavior**
No errors.

**VS Code extension or command-line**
Pylance language server 2021.7.4 (pyright 209f9621)
pyright 1.1.157

**Additional context**
If the signature of `b` is changed to `b(self, v: A[U]) -> A[Tuple[Tuple[T, U], U]]` in both the protocol and the class, Pyright stops complaining but starts to accept code that fails at runtime with a `TypeError`:

```python
def c(x: Tuple[Tuple[int, str], str]) -> int:
    return x[0][0]


reveal_type(B(1).b(B('a')).a())  # Type of "B(1).b(B('a')).a()" is "Tuple[Tuple[int, str], str]"
c(B(1).b(B('a')).a())  # TypeError: 'int' object is not subscriptable
```

**Comment by erictraut** [comment]
Here's what's happening. For purposes of explanation, I'm going to simplify the sample and remove all references to `U` and `a`. The bug isn't related to those, and it's easier to see what's happening if they're removed.

```python
T = TypeVar("T")

class A(Protocol[T]):
    def b(self) -> "A[Tuple[T]]":
        ...

class B(Generic[T]):
    def __init__(self, t: T):
        self._t = t

    def b(self) -> A[Tuple[T]]:
        x = (self._t,)
       reveal_type(x) # Tuple[T@B]
       y = B(x)
       reveal_type(y) # B[Tuple[T@B]]
       return y
```

The type of `x` is correctly inferred to be `Tuple[T@B]`. When this is passed to the constructor of `B`, it results in a value `y` that is of type `B[Tuple[T@B]]`. So far, so good.

Next, the type analyzer needs to determine whether `B[Tuple[T@B]]` is compatible with the declared return type `A[Tuple[T@B]]`. Since `A` is a protocol class with no explicit relation to `B` (i.e. `B` doesn't explicitly derive from `A`), it needs to do a protocol comparison. To do so, it needs to first specialize `A` with the provided type argument — effectively replacing all instances of type parameter `T@A` with the type argument `Tuple[T@B]`. The unspecialized version of method `A.b` is `(self) -> A[Tuple[T@A]]`. After specialization, its type becomes `(self) -> A[Tuple[Tuple[T@B]]`.

Next, the protocol comparison logic compares each symbol defined in the specialized protocol `A` with the corresponding symbol defined in `B`. When it compares method `A.b` with `B.b`, it's asking whether type `(self) -> A[Tuple[T@B]]` is compatible with `(self) -> A[Tuple[Tuple[T@B]]`. Using normal type compatibility rules, the answer is no, these types are incompatible. The following error message is the result:

```
Expression of type "B[tuple[T@B]]" cannot be assigned to return type "A[Tuple[T@B]]"
  TypeVar "T@A" is invariant
    Tuple entry 1 is incorrect type
      Type "tuple[T@B]" cannot be assigned to type "T@B"
```

Intuitively, it seems like this code should type check without errors, and mypy accepts it, but I'm still not seeing the fault in the logic above. I'll keep thinking about it.

There are two simple workarounds I can offer:
1. Change the declaration of `B` so it explicitly derives from `A`.

```python
class B(A[T]):
   ...
```

2. Change the return type of method `B.b` to `B[Tuple[T, U]]`.

**Comment by ethframe** [comment]
I've reduced the sample even more:

```python
from typing import Generic, Tuple, TypeVar

T = TypeVar("T")


class A(Generic[T]):
    def __init__(self, t: T):
        self._t = t

    def b(self) -> 'A[Tuple[T]]':
        y = A((self._t,))
        reveal_type(y)  # A[tuple[T@A]]
        reveal_type(y.b)  # () -> A[Tuple[tuple[tuple[T@A]]]]
        return y


def b(t: T) -> A[Tuple[T]]:
    y = A(t).b()
    reveal_type(y)  # A[Tuple[T@b]]
    reveal_type(y.b)  # () -> A[Tuple[Tuple[T@b]]]
    return y
```

The type of `y.b` inside `A.b` is incorrect.

**Comment by erictraut** [comment]
Thanks for the simplified sample. This turned out to be a slightly different problem, one that was straightforward to fix. With the latest version of pyright (1.1.161), your simplified sample above now type checks correctly. But the original sample still doesn't.

**Comment by ethframe** [comment]
Tried 1.1.161, still getting the same error even for a simplified sample. And I suppose that the original sample does not type checks because of the double specialization bug:

```python
from typing import Generic, Tuple, TypeVar

from typing_extensions import Protocol

T = TypeVar("T")


class A(Protocol[T]):
    def a(self) -> 'A[Tuple[T]]': ...


class B(Generic[T]):
    def __init__(self, x: T):
        self.x = x

    def a(self) -> A[Tuple[T]]:
        t = (self.x,)
        y = B(t)
        reveal_type(y.a)  # () -> A[Tuple[tuple[tuple[T@B]]]]
        return y  # Error

# Hack to get differently scoped T inside method body:

def a(self: 'C[T]') -> A[Tuple[T]]:
    t = (self.x,)
    y = C(t)
    reveal_type(y.a)  # () -> A[Tuple[tuple[T@a]]]
    return y  # No error


class C(Generic[T]):
    def __init__(self, x: T):
        self.x = x

    a = a
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.174, which I just published. It will also be included in the next release of pylance.

**Comment by erictraut** [comment]
Unfortunately, I needed to revert the fix for this issue because it broke some things. I'll need to continue to think about a better fix. 

The problem is a bad interaction with the mechanism pyright uses for unannotated `self` and `cls` parameters.

**Comment by erictraut** [comment]
I think I finally found a good fix for this bug. The fix will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.179, which I just published. It will also be included in the next release of pylance.

**Comment by ethframe** [comment]
The bug seems to be partially fixed. While the sample from https://github.com/microsoft/pyright/issues/2103#issuecomment-895798533 now passes typechecking, the inferred type for `y.a` is still `() -> A[Tuple[tuple[tuple[T@B]]]]`, instead of `() -> A[Tuple[tuple[T@B]]]`. Adding an explicit annotation for `self` fixes this:

```python
from typing import Generic, Tuple, TypeVar

from typing_extensions import Protocol

T = TypeVar("T")


class A(Protocol[T]):
    def a(self) -> 'A[Tuple[T]]': ...


class B(Generic[T]):
    def __init__(self, x: T):
        self.x = x

    def a(self) -> A[Tuple[T]]:
        t = (self.x,)
        y = B(t)
        reveal_type(y)  # B[Tuple[T@B]]
        reveal_type(y.a())  # A[Tuple[tuple[tuple[T@B]]]]
        return y

    def b(self: 'B[T]') -> A[Tuple[T]]:
        t = (self.x,)
        y = B(t)
        reveal_type(y)  # B[Tuple[T@B]]
        reveal_type(y.b())  # A[Tuple[tuple[T@B]]]
        return y
```

**Comment by ethframe** [comment]
I've come up with a better example of the problem. It is still possible to trick pyright into believing that some value is of type `tuple[tuple[T]]`, while in runtime the value is `tuple[T]`:

```python
from typing import Generic, Tuple, TypeVar

from typing_extensions import Protocol

T = TypeVar("T")


class A(Protocol[T]):
    def a(self) -> 'A[Tuple[T]]': ...
    def b(self) -> 'A[Tuple[T]]': ...
    def c(self) -> 'T': ...


class B(Generic[T]):
    def __init__(self, x: T):
        self.x = x

    def a(self) -> A[Tuple[T]]:
        t = (self.x,)
        y = B(t)
        v = f(y.b())
        reveal_type(v)  # tuple[tuple[T@B]]
        reveal_type(v[0][0])  # T@B
        v[0][0]  # Passes type checking, but raises exception at runtime
        return y

    def b(self) -> A[Tuple[T]]:
        t = (self.x,)
        y = B(t)
        return y

    def c(self) -> T:
        return self.x


def f(a: A[Tuple[T]]) -> T:
    return a.c()[0]


B(1).a()  # TypeError: 'int' object is not subscriptable
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.180, which I just published. It will also be included in the next pylance release.
